New version of NetAnalyzers throws CA1420 for delegates

### Analyzer

**Diagnostic ID**: CA1420 `Delegates with managed types as parameters or the return type require runtime marshalling to be enabled in the assembly where the delegate is defined`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 7.0.0-preview1.22302.1

### Describe the bug

D:\a\_work\1\s\src\Controls\src\Core\Platform\Tizen\Extensions\DragDropExtensions.cs(89,28): error CA1420: Delegates with managed types as parameters or the return type require runtime marshalling to be enabled in the assembly where the delegate is defined. [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj]
  
I started getting this issue after updating the package from:

 - GOOD: 7.0.0-preview1.22274.2
 - BAD: 7.0.0-preview1.22302.1
 - BAD: 7.0.0-preview1.22379.1


### Steps To Reproduce

1. Clone: https://github.com/dotnet/maui/tree/release/7.0.1xx-preview7
2. Update https://github.com/dotnet/maui/blob/release/7.0.1xx-preview7/eng/Versions.props#L39 to be the newer: 7.0.0-preview1.22302.1
3. `dotnet build Microsoft.Maui.BuildTasks.slnf /p:configuration=Release`

### Expected behavior
Builds.

### Actual behavior
Error.

## Additional context

You may need to install the tizen workload, but I also saw this on other projects. I know this is a vague issue with a "download my code and build" but I am not sure what this error even means. Is it saying that delegates are not good?

I _think_ it seems to be if some p/invoke has a delegate that uses a IntPtr

Hi,
Another reproduction:
```powershell
dotnet new console
```
Replace the content of "Program.cs" with:
```cs
using System.Runtime.InteropServices;

class Program
{
    [StructLayout(LayoutKind.Sequential, Pack = 1, CharSet = CharSet.Auto)]
    public struct testStruct
    {
        public CULong testPrp;
    }

    [UnmanagedFunctionPointerAttribute(CallingConvention.StdCall, CharSet = CharSet.Auto)]
    internal delegate CULong fncDlg(ref testStruct strcPrm);
    static void Main()
    {
        var lb = NativeLibrary.Load(@"test.dll");
        var fnc = NativeLibrary.GetExport(lb, "testFnc");

        testStruct strcObj = new testStruct();

        //#pragma warning disable CA1420
        var dlg = Marshal.GetDelegateForFunctionPointer<fncDlg>(fnc);
        //#pragma warning restore CA1420

        dlg(ref strcObj);

        NativeLibrary.Free(lb);
    }
}
```
and
```powershell
dotnet build
```
show:
```
Program.cs(21,19): warning CA1420: Delegates with managed types as parameters or the return type require runtime marshalling to be enabled in the assembly where the delegat 
e is defined.
```
dotnet version: 7.0.100-preview.7.22377.5

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs
@@ -149,7 +149,7 @@ namespace Microsoft.NetCore.Analyzers.InteropServices
                         var disabledRuntimeMarshallingAssemblyAnalyzer = new DisabledRuntimeMarshallingAssemblyAnalyzer(context.Compilation, autoLayoutCache);
                         disabledRuntimeMarshallingAssemblyAnalyzer.RegisterActions(context);
                     }
-                    var delegateInteropUsageAnalyzer = new DelegateInteropUsageAnalyzer(context.Compilation, autoLayoutCache);
+                    var delegateInteropUsageAnalyzer = new DelegateInteropUsageAnalyzer(context.Compilation, autoLayoutCache, disableRuntimeMarshallingAttribute);
                     delegateInteropUsageAnalyzer.RegisterActions(context, hasDisableRuntimeMarshallingAttribute);
                 }
             });

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshallingAnalyzer.DelegateInteropUsageAnalyzer.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshallingAnalyzer.DelegateInteropUsageAnalyzer.cs
@@ -20,8 +20,9 @@ namespace Microsoft.NetCore.Analyzers.InteropServices
             private readonly IMethodSymbol? _getFunctionPointerForDelegateNonGeneric;
             private readonly IMethodSymbol? _getFunctionPointerForDelegateGeneric;
             private readonly AutoLayoutTypeCache _autoLayoutCache;
+            private readonly INamedTypeSymbol _disableRuntimeMarshallingAttribute;
 
-            public DelegateInteropUsageAnalyzer(Compilation compilation, AutoLayoutTypeCache autoLayoutCache)
+            public DelegateInteropUsageAnalyzer(Compilation compilation, AutoLayoutTypeCache autoLayoutCache, INamedTypeSymbol disableRuntimeMarshallingAttribute)
             {
                 if (compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesMarshal, out var marshalType))
                 {
@@ -34,6 +35,7 @@ namespace Microsoft.NetCore.Analyzers.InteropServices
                 }
 
                 _autoLayoutCache = autoLayoutCache;
+                _disableRuntimeMarshallingAttribute = disableRuntimeMarshallingAttribute;
             }
 
             public void RegisterActions(CompilationStartAnalysisContext context, bool hasDisableRuntimeMarshallingAttribute)
@@ -88,7 +90,10 @@ namespace Microsoft.NetCore.Analyzers.InteropServices
 
                 void AnalyzeDelegateType(OperationAnalysisContext context, IOperation operation, INamedTypeSymbol delegateType)
                 {
-                    AnalyzeMethodSignature(_autoLayoutCache, context.ReportDiagnostic, delegateType.DelegateInvokeMethod, ImmutableArray.Create(operation.Syntax.GetLocation()), FeatureUnsupportedWhenRuntimeMarshallingDisabledDelegateUsage);
+                    if (delegateType.ContainingAssembly.HasAttribute(_disableRuntimeMarshallingAttribute))
+                    {
+                        AnalyzeMethodSignature(_autoLayoutCache, context.ReportDiagnostic, delegateType.DelegateInvokeMethod, ImmutableArray.Create(operation.Syntax.GetLocation()), FeatureUnsupportedWhenRuntimeMarshallingDisabledDelegateUsage);
+                    }
                 }
             }
 
@@ -132,7 +137,10 @@ namespace Microsoft.NetCore.Analyzers.InteropServices
 
                 void AnalyzeDelegateMethodSignature(INamedTypeSymbol delegateType, ISymbol signatureSymbol)
                 {
-                    AnalyzeMethodSignature(_autoLayoutCache, reportDiagnostic, delegateType.DelegateInvokeMethod, signatureSymbol.Locations, FeatureUnsupportedWhenRuntimeMarshallingDisabledDelegateUsage);
+                    if (delegateType.ContainingAssembly.HasAttribute(_disableRuntimeMarshallingAttribute))
+                    {
+                        AnalyzeMethodSignature(_autoLayoutCache, reportDiagnostic, delegateType.DelegateInvokeMethod, signatureSymbol.Locations, FeatureUnsupportedWhenRuntimeMarshallingDisabledDelegateUsage);
+                    }
                 }
             }
         }

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshallingTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshallingTests.cs
@@ -804,6 +804,55 @@ class C
         }
 
         [Fact]
+        [WorkItem(6094, "https://github.com/dotnet/roslyn-analyzers/issues/6094")]
+        public async Task CS_DelegateWith_NonBlittableParameter_PInvokeUsage_NoDisableRuntimeMarshalling_DoesNotEmit_Diagnostic()
+        {
+            string source = @"
+using System;
+using System.Runtime.InteropServices;
+
+internal delegate void DelegateType(string[] str);
+
+class C
+{
+    [DllImport(""Native"")]
+    public static extern void PInvoke(DelegateType d);
+    [DllImport(""Native"")]
+    public static extern DelegateType PInvoke();
+    [DllImport(""Native"")]
+    public static extern void PInvokeArray(DelegateType[] d);
+    [DllImport(""Native"")]
+    public static extern DelegateType[] PInvokeArray();
+}
+";
+            await VerifyCSAnalyzerAsync(source);
+        }
+
+        [Fact]
+        [WorkItem(6094, "https://github.com/dotnet/roslyn-analyzers/issues/5995")]
+        public async Task CS_DelegateWithClassReturnValue_NoUnmanagedFunctionPointer_MarshalAPIUsage_NoDisableRuntimeMarshalling_DoesNotEmit_Diagnostic()
+        {
+            string source = @"
+using System;
+using System.Runtime.InteropServices;
+
+public delegate string DelegateType();
+
+class C
+{
+    public static void M(DelegateType d, IntPtr i)
+    {
+        _ = Marshal.GetFunctionPointerForDelegate((Delegate)d);
+        _ = Marshal.GetFunctionPointerForDelegate(d);
+        _ = Marshal.GetDelegateForFunctionPointer(i, typeof(DelegateType));
+        _ = Marshal.GetDelegateForFunctionPointer<DelegateType>(i);
+    }
+}
+";
+            await VerifyCSAnalyzerAsync(source);
+        }
+
+        [Fact]
         public async Task VB_DelegateWithClassReturnValue_Emits_Diagnostic()
         {
             string source = @"

